extension/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+    step_size: TimestampTz,
+    greatest_time: TimestampTz,
+    is_counter: bool,
+    is_rate: bool,
+}
+
+impl GapfillDeltaTransition {
+    pub fn new(
+        lowest_time: TimestampTz,
+        greatest_time: TimestampTz,
+        window_size: Milliseconds,
+        step_size: Milliseconds,
+        is_counter: bool,
+        is_rate: bool
+    ) -> Self {
+        let mut expected_deltas = (greatest_time - lowest_time) / (step_size * USECS_PER_MS);
+        if (greatest_time - lowest_time) % (step_size * USECS_PER_MS) != 0 {
+            expected_deltas += 1
+        }
+        GapfillDeltaTransition{
+            window: VecDeque::default(),
+            deltas: Vec::with_capacity(expected_deltas as usize),
+            nulls: Vec::with_capacity(expected_deltas as usize),
+            current_window_max: lowest_time + window_size*USECS_PER_MS,
+            current_window_min: lowest_time,
+            step_size: step_size*USECS_PER_MS,
+            greatest_time,
+            is_counter,
+            is_rate,
+        }
+    }
+
+    fn add_data_point(&mut self, time: TimestampTz, val: f64) {
+        // skip stale NaNs
+        const STALE_NAN: u64 = 0x7ff0000000000002;
+        if val.to_bits() == STALE_NAN {
+            return
+        }
+
+        while !self.in_current_window(time) {
+            self.flush_current_window()
+        }
+
+        self.window.push_back((time, val));
+    }
+
+    fn in_current_window(&self, time: TimestampTz) -> bool {
+        time <= self.current_window_max
+    }
+
+    fn flush_current_window(&mut self) {
+        self.add_delta_for_current_window();
+
+        self.current_window_min += self.step_size;
+        self.current_window_max += self.step_size;
+
+        let current_window_min = self.current_window_min;
+        while self.window.front().map(|(time, _)| *time < current_window_min).unwrap_or(false) {
+            self.window.pop_front();
+        }
+    }
+
+    //based on extrapolatedRate
+    // https://github.com/prometheus/prometheus/blob/e5ffa8c9a08a5ee4185271c8c26051ddc1388b7a/promql/functions.go#L59
+    fn add_delta_for_current_window(&mut self) {
+        if self.window.len() < 2 {
+            // if there are 1 or fewer values in the window, store NULL
+            self.nulls.push(true);
+            return
+        }
+
+        let mut counter_correction = 0.0;
+        if self.is_counter {
+            let mut last_value = 0.0;
+            for (_, sample) in &self.window {
+                if *sample < last_value {
+                    counter_correction += last_value
+                }
+                last_value = *sample
+            }
+        }
+
+        let (latest_time, latest_val)= self.window.back().cloned().unwrap();
+        let (earliest_time, earliest_val) = self.window.front().cloned().unwrap();
+        let mut result_val = latest_val - earliest_val + counter_correction;
+
+        let mut duration_to_start = (earliest_time - self.current_window_min) as f64
```

Comment:
can we indicate this is seconds either in comments or through naming convention

Reply:
comment I think

Reply:
or we can use newtypes if you really want